File: backend/app/institution_modules.py
"""Canonical product modules and per-role staff sidebar access."""

from __future__ import annotations

from typing import Any

ALL_MODULE_IDS = (
    "enrollment",
    "scholarships",
    "support",
    "grants",
    "rankings",
)

STUDENT_MODULE_IDS = (
    "enrollment",
    "scholarships",
    "support",
    "grants",
)

STAFF_MODULE_IDS = ALL_MODULE_IDS

STAFF_ACCESS_ROLES = (
    "vice_chancellor",
    "registrar",
    "scholarship_office",
    "student_services",
    "research_office",
)

# Controllable staff sidebar items per module (paths used for grants + route guards)
STAFF_MODULE_NAV_ITEMS: dict[str, list[dict[str, Any]]] = {
    "enrollment": [
        {"id": "at_risk", "path": "/staff/at-risk"},
        {
            "id": "students",
            "path": "/staff/students",
            "match_prefixes": ["/staff/enrollment"],
        },
    ],
    "support": [
        {"id": "support", "path": "/staff/support"},
    ],
    "scholarships": [
        {"id": "scholarships", "path": "/staff/scholarships"},
        {"id": "financial_aid", "path": "/staff/financial-aid"},
        {"id": "applications", "path": "/staff/scholarships/applications"},
        {"id": "triage", "path": "/staff/scholarships/triage"},
        {"id": "decisions", "path": "/staff/scholarships/decisions"},
        {"id": "opportunities", "path": "/staff/scholarships/opportunities"},
        {"id": "configure", "path": "/staff/scholarships/configure"},
    ],
    "grants": [
        {"id": "grants", "path": "/staff/grants"},
        {"id": "lifecycle", "path": "/staff/grants/lifecycle"},
        {"id": "applications", "path": "/staff/grants/applications"},
        {"id": "opportunities", "path": "/staff/grants/opportunities"},
        {"id": "configure", "path": "/staff/grants/configure"},
    ],
    "rankings": [
        {"id": "rankings", "path": "/staff/rankings"},
    ],
}

FULL_MODULE = "*"

DEFAULT_ENABLED_MODULES = {
    "student": list(STUDENT_MODULE_IDS),
    "staff": list(STAFF_MODULE_IDS),
}

_ALLOWED_BY_PORTAL = {
    "student": set(STUDENT_MODULE_IDS),
    "staff": set(STAFF_MODULE_IDS),
}


def default_enabled_modules() -> dict[str, list[str]]:
    return {
        "student": list(STUDENT_MODULE_IDS),
        "staff": list(STAFF_MODULE_IDS),
    }


def normalize_enabled_modules(raw: Any) -> dict[str, list[str]]:
    """Treat missing/legacy null as all modules enabled."""
    if not isinstance(raw, dict):
        return default_enabled_modules()
    out = default_enabled_modules()
    for portal, allowed in _ALLOWED_BY_PORTAL.items():
        items = raw.get(portal)
        if not isinstance(items, list):
            continue
        seen: list[str] = []
        for item in items:
            if isinstance(item, str) and item in allowed and item not in seen:
                seen.append(item)
        out[portal] = seen
    return out


def is_module_enabled(raw: Any, portal: str, module_id: str) -> bool:
    enabled = normalize_enabled_modules(raw)
    return module_id in enabled.get(portal, [])


def _role_value(role: Any) -> str:
    if hasattr(role, "value"):
        return str(role.value)
    return str(role or "")


def module_item_paths(module_id: str) -> list[str]:
    return [str(item["path"]) for item in STAFF_MODULE_NAV_ITEMS.get(module_id, [])]


def _path_matches_item(pathname: str, item: dict[str, Any]) -> bool:
    path = str(item.get("path") or "")
    if not path or not pathname:
        return False
    if pathname == path or pathname.startswith(f"{path}/"):
        return True
    for prefix in item.get("match_prefixes") or []:
        prefix = str(prefix)
        if pathname == prefix or pathname.startswith(f"{prefix}/"):
            return True
    return False


def find_staff_nav_item(pathname: str) -> dict[str, Any] | None:
    """Most specific catalog item for a staff path."""
    matches: list[tuple[int, str, dict[str, Any]]] = []
    for module_id, items in STAFF_MODULE_NAV_ITEMS.items():
        for item in items:
            if _path_matches_item(pathname, item):
                matches.append((len(str(item["path"])), module_id, item))
    if not matches:
        return None
    matches.sort(key=lambda row: row[0], reverse=True)
    _, module_id, item = matches[0]
    return {"module": module_id, "item": item}


def _normalize_module_grant(value: Any, module_id: str) -> Any:
    paths = module_item_paths(module_id)
    path_set = set(paths)
    if value == FULL_MODULE or value == "all" or value is True:
        return FULL_MODULE
    if isinstance(value, list):
        seen: list[str] = []
        for entry in value:
            if isinstance(entry, str) and entry in path_set and entry not in seen:
                seen.append(entry)
        # Preserve catalog order
        ordered = [p for p in paths if p in set(seen)]
        if not ordered:
            return []
        if len(ordered) == len(paths):
            return FULL_MODULE
        return ordered
    return []


def _legacy_modules_list_to_grants(modules: list[str], ceiling: list[str]) -> dict[str, Any]:
    ceiling_set = set(ceiling)
    out: dict[str, Any] = {}
    for module_id in STAFF_MODULE_IDS:
        if module_id in ceiling_set and module_id in modules:
            out[module_id] = FULL_MODULE
        else:
            out[module_id] = []
    return out


def default_role_grants(ceiling: list[str]) -> dict[str, Any]:
    ceiling_set = set(ceiling)
    return {
        module_id: (FULL_MODULE if module_id in ceiling_set else [])
        for module_id in STAFF_MODULE_IDS
    }


def normalize_staff_role_modules(
    raw: Any,
    enabled_staff: list[str] | None = None,
) -> dict[str, dict[str, Any]]:
    """
    Per-role grants:
      { role: { moduleId: "*" | [paths...] } }

    Legacy shape { role: ["enrollment", "rankings"] } means full access to those modules.
    Missing role keys default to full access for every institution-enabled module.
    """
    ceiling = list(enabled_staff) if enabled_staff is not None else list(STAFF_MODULE_IDS)
    ceiling = [m for m in STAFF_MODULE_IDS if m in set(ceiling)]
    out = {role: default_role_grants(ceiling) for role in STAFF_ACCESS_ROLES}
    if not isinstance(raw, dict):
        return out

    for role in STAFF_ACCESS_ROLES:
        entry = raw.get(role)
        if entry is None:
            continue
        if isinstance(entry, list):
            # Legacy: list of module ids
            modules = [m for m in entry if isinstance(m, str)]
            grants = _legacy_modules_list_to_grants(modules, ceiling)
        elif isinstance(entry, dict):
            grants = default_role_grants([])  # start empty, then fill
            for module_id in STAFF_MODULE_IDS:
                if module_id not in ceiling:
                    grants[module_id] = []
                    continue
                if module_id not in entry:
                    # Unspecified module in new format → no access (explicit grants only)
                    # But if the dict is empty, treat as default full? Prefer: missing key = full when
                    # migrating from partial saves. Safer UX: missing key keeps previous default full
                    # only when role entry was absent entirely. Here entry exists → missing module = off
                    grants[module_id] = _normalize_module_grant(entry.get(module_id, []), module_id)
                else:
                    grants[module_id] = _normalize_module_grant(entry[module_id], module_id)
            # Clip to ceiling
            for module_id in STAFF_MODULE_IDS:
                if module_id not in ceiling:
                    grants[module_id] = []
        else:
            continue
        out[role] = grants
    return out


def modules_from_role_grants(grants: dict[str, Any] | None) -> list[str]:
    if not isinstance(grants, dict):
        return []
    enabled: list[str] = []
    for module_id in STAFF_MODULE_IDS:
        value = grants.get(module_id)
        if value == FULL_MODULE:
            enabled.append(module_id)
        elif isinstance(value, list) and value:
            enabled.append(module_id)
    return enabled


def role_allows_staff_path(grants: dict[str, Any] | None, pathname: str) -> bool:
    """True if grants allow this staff path (or path is outside the catalog)."""
    match = find_staff_nav_item(pathname)
    if not match:
        return True
    if not isinstance(grants, dict):
        return True
    module_id = match["module"]
    item_path = str(match["item"]["path"])
    value = grants.get(module_id)
    if value == FULL_MODULE:
        return True
    if isinstance(value, list):
        return item_path in value
    return False


def effective_enabled_modules_for_user(
    institution_modules: Any,
    staff_role_modules: Any,
    role: Any,
    account_category: str | None = None,
) -> dict[str, list[str]]:
    """Institution modules intersected with this user's role grants (staff roles only)."""
    base = normalize_enabled_modules(institution_modules)
    role_key = _role_value(role)
    if role_key not in STAFF_ACCESS_ROLES:
        return base
    role_map = normalize_staff_role_modules(staff_role_modules, base["staff"])
    return {
        "student": list(base["student"]),
        "staff": modules_from_role_grants(role_map.get(role_key)),
    }


def effective_staff_role_access_for_user(
    institution_modules: Any,
    staff_role_modules: Any,
    role: Any,
) -> dict[str, Any] | None:
    """Granular grants for the current staff role, or None when not a configurable role."""
    role_key = _role_value(role)
    if role_key not in STAFF_ACCESS_ROLES:
        return None
    base = normalize_enabled_modules(institution_modules)
    role_map = normalize_staff_role_modules(staff_role_modules, base["staff"])
    return role_map.get(role_key)

File: frontend/src/lib/institutionModules.js
export const INSTITUTION_MODULES = [
  { id: "enrollment", student: true, staff: true },
  { id: "scholarships", student: true, staff: true },
  { id: "support", student: true, staff: true },
  { id: "grants", student: true, staff: true },
  { id: "rankings", student: false, staff: true },
];

export const DEFAULT_ENABLED_MODULES = {
  student: ["enrollment", "scholarships", "support", "grants"],
  staff: ["enrollment", "scholarships", "support", "grants", "rankings"],
};

export const STAFF_ACCESS_ROLES = [
  "vice_chancellor",
  "registrar",
  "scholarship_office",
  "student_services",
  "research_office",
];

/** Controllable staff sidebar items grouped by product module */
export const STAFF_MODULE_NAV_ITEMS = {
  enrollment: [
    { id: "at_risk", path: "/staff/at-risk", labelKey: "atRisk" },
    {
      id: "students",
      path: "/staff/students",
      labelKey: "students",
      matchPrefixes: ["/staff/enrollment"],
    },
  ],
  support: [{ id: "support", path: "/staff/support", labelKey: "support" }],
  scholarships: [
    { id: "scholarships", path: "/staff/scholarships", labelKey: "scholarships" },
    { id: "financial_aid", path: "/staff/financial-aid", labelKey: "financialAid" },
    { id: "applications", path: "/staff/scholarships/applications", labelKey: "applications" },
    { id: "triage", path: "/staff/scholarships/triage", labelKey: "triage" },
    { id: "decisions", path: "/staff/scholarships/decisions", labelKey: "decisions" },
    { id: "opportunities", path: "/staff/scholarships/opportunities", labelKey: "scholarshipOpportunities" },
    { id: "configure", path: "/staff/scholarships/configure", labelKey: "configureScholarships" },
  ],
  grants: [
    { id: "grants", path: "/staff/grants", labelKey: "grants" },
    { id: "lifecycle", path: "/staff/grants/lifecycle", labelKey: "grantLifecycle" },
    { id: "applications", path: "/staff/grants/applications", labelKey: "grantApplications" },
    { id: "opportunities", path: "/staff/grants/opportunities", labelKey: "grantOpportunities" },
    { id: "configure", path: "/staff/grants/configure", labelKey: "configureGrants" },
  ],
  rankings: [{ id: "rankings", path: "/staff/rankings", labelKey: "rankings" }],
};

export const FULL_MODULE = "*";

const STUDENT_ALWAYS = ["/student", "/student/profile"];
const STAFF_ALWAYS = ["/staff", "/staff/profile", "/staff/analytics"];
const INSTITUTION_ADMIN_ALWAYS = [
  "/institution/admin",
  "/institution/admin/analytics",
  "/institution/admin/users",
  "/institution/admin/domains",
  "/institution/admin/profile",
  "/institution/admin/access",
  "/institution/admin/activity",
];

const STUDENT_PATH_MODULES = [
  { prefixes: ["/student/courses", "/student/grades", "/student/attendance"], modules: ["enrollment"] },
  { prefixes: ["/student/scholarships"], modules: ["scholarships"] },
  { prefixes: ["/student/grants"], modules: ["grants"] },
  { prefixes: ["/student/financials"], modules: ["scholarships", "grants"] },
  { prefixes: ["/student/support"], modules: ["support"] },
];

const STAFF_PATH_MODULES = [
  { prefixes: ["/staff/students", "/staff/enrollment", "/staff/at-risk"], modules: ["enrollment"] },
  { prefixes: ["/staff/scholarships", "/staff/financial-aid"], modules: ["scholarships"] },
  { prefixes: ["/staff/support"], modules: ["support"] },
  { prefixes: ["/staff/grants"], modules: ["grants"] },
  { prefixes: ["/staff/rankings"], modules: ["rankings"] },
];

const INSTITUTION_ADMIN_PATH_MODULES = [
  { prefixes: ["/institution/admin/grants"], modules: ["grants"] },
];

export function normalizeEnabledModules(raw) {
  const fallback = {
    student: [...DEFAULT_ENABLED_MODULES.student],
    staff: [...DEFAULT_ENABLED_MODULES.staff],
  };
  if (!raw || typeof raw !== "object") return fallback;
  const allowed = {
    student: new Set(DEFAULT_ENABLED_MODULES.student),
    staff: new Set(DEFAULT_ENABLED_MODULES.staff),
  };
  const out = { ...fallback };
  for (const portal of ["student", "staff"]) {
    const items = raw[portal];
    if (!Array.isArray(items)) continue;
    const seen = [];
    for (const item of items) {
      if (typeof item === "string" && allowed[portal].has(item) && !seen.includes(item)) {
        seen.push(item);
      }
    }
    out[portal] = seen;
  }
  return out;
}

export function moduleItemPaths(moduleId) {
  return (STAFF_MODULE_NAV_ITEMS[moduleId] || []).map((item) => item.path);
}

function pathMatchesItem(pathname, item) {
  if (!pathname || !item?.path) return false;
  if (pathname === item.path || pathname.startsWith(`${item.path}/`)) return true;
  return (item.matchPrefixes || []).some(
    (prefix) => pathname === prefix || pathname.startsWith(`${prefix}/`)
  );
}

export function findStaffNavCatalogItem(pathname) {
  const matches = [];
  for (const [moduleId, items] of Object.entries(STAFF_MODULE_NAV_ITEMS)) {
    for (const item of items) {
      if (pathMatchesItem(pathname, item)) {
        matches.push({ moduleId, item, score: item.path.length });
      }
    }
  }
  if (!matches.length) return null;
  matches.sort((a, b) => b.score - a.score);
  return matches[0];
}

function normalizeModuleGrant(value, moduleId) {
  const paths = moduleItemPaths(moduleId);
  const pathSet = new Set(paths);
  if (value === FULL_MODULE || value === "all" || value === true) return FULL_MODULE;
  if (!Array.isArray(value)) return [];
  const seen = [];
  for (const entry of value) {
    if (typeof entry === "string" && pathSet.has(entry) && !seen.includes(entry)) {
      seen.push(entry);
    }
  }
  const ordered = paths.filter((p) => seen.includes(p));
  if (!ordered.length) return [];
  if (ordered.length === paths.length) return FULL_MODULE;
  return ordered;
}

function defaultRoleGrants(ceiling) {
  const ceilingSet = new Set(ceiling);
  return Object.fromEntries(
    DEFAULT_ENABLED_MODULES.staff.map((moduleId) => [
      moduleId,
      ceilingSet.has(moduleId) ? FULL_MODULE : [],
    ])
  );
}

function legacyModulesToGrants(modules, ceiling) {
  const ceilingSet = new Set(ceiling);
  const selected = new Set(modules);
  return Object.fromEntries(
    DEFAULT_ENABLED_MODULES.staff.map((moduleId) => [
      moduleId,
      ceilingSet.has(moduleId) && selected.has(moduleId) ? FULL_MODULE : [],
    ])
  );
}

export function normalizeStaffRoleModules(raw, enabledStaff = null) {
  const ceiling = Array.isArray(enabledStaff)
    ? DEFAULT_ENABLED_MODULES.staff.filter((id) => enabledStaff.includes(id))
    : [...DEFAULT_ENABLED_MODULES.staff];
  const out = Object.fromEntries(
    STAFF_ACCESS_ROLES.map((role) => [role, defaultRoleGrants(ceiling)])
  );
  if (!raw || typeof raw !== "object") return out;

  for (const role of STAFF_ACCESS_ROLES) {
    const entry = raw[role];
    if (entry == null) continue;
    if (Array.isArray(entry)) {
      out[role] = legacyModulesToGrants(
        entry.filter((m) => typeof m === "string"),
        ceiling
      );
      continue;
    }
    if (typeof entry !== "object") continue;
    const grants = defaultRoleGrants([]);
    for (const moduleId of DEFAULT_ENABLED_MODULES.staff) {
      if (!ceiling.includes(moduleId)) {
        grants[moduleId] = [];
        continue;
      }
      grants[moduleId] = normalizeModuleGrant(
        Object.prototype.hasOwnProperty.call(entry, moduleId) ? entry[moduleId] : [],
        moduleId
      );
    }
    out[role] = grants;
  }
  return out;
}

export function modulesFromRoleGrants(grants) {
  if (!grants || typeof grants !== "object") return [];
  return DEFAULT_ENABLED_MODULES.staff.filter((moduleId) => {
    const value = grants[moduleId];
    return value === FULL_MODULE || (Array.isArray(value) && value.length > 0);
  });
}

export function isModuleEnabled(raw, portal, moduleId) {
  return normalizeEnabledModules(raw)[portal]?.includes(moduleId) ?? false;
}

export function isRoleModuleFullyEnabled(grants, moduleId) {
  return grants?.[moduleId] === FULL_MODULE;
}

export function isRoleNavItemEnabled(grants, moduleId, path) {
  const value = grants?.[moduleId];
  if (value === FULL_MODULE) return true;
  if (Array.isArray(value)) return value.includes(path);
  return false;
}

export function roleModuleSelectionState(grants, moduleId) {
  const paths = moduleItemPaths(moduleId);
  const value = grants?.[moduleId];
  if (value === FULL_MODULE) return { checked: true, indeterminate: false, selected: paths };
  if (!Array.isArray(value) || !value.length) {
    return { checked: false, indeterminate: false, selected: [] };
  }
  return {
    checked: false,
    indeterminate: true,
    selected: value,
  };
}

/** Toggle whole module for a role (full ↔ off). */
export function toggleRoleModuleSelection(staffRoleModules, role, moduleId, enabledModules = null) {
  const ceiling = normalizeEnabledModules(enabledModules).staff;
  const current = normalizeStaffRoleModules(staffRoleModules, ceiling);
  if (!STAFF_ACCESS_ROLES.includes(role) || !ceiling.includes(moduleId)) return current;
  const nextGrant = current[role][moduleId] === FULL_MODULE ? [] : FULL_MODULE;
  return {
    ...current,
    [role]: {
      ...current[role],
      [moduleId]: nextGrant,
    },
  };
}

/** Toggle a single sidebar item within a module. */
export function toggleRoleNavItemSelection(
  staffRoleModules,
  role,
  moduleId,
  path,
  enabledModules = null
) {
  const ceiling = normalizeEnabledModules(enabledModules).staff;
  const current = normalizeStaffRoleModules(staffRoleModules, ceiling);
  if (!STAFF_ACCESS_ROLES.includes(role) || !ceiling.includes(moduleId)) return current;
  const paths = moduleItemPaths(moduleId);
  if (!paths.includes(path)) return current;

  let selected;
  const value = current[role][moduleId];
  if (value === FULL_MODULE) selected = [...paths];
  else if (Array.isArray(value)) selected = [...value];
  else selected = [];

  if (selected.includes(path)) selected = selected.filter((p) => p !== path);
  else selected.push(path);

  const ordered = paths.filter((p) => selected.includes(p));
  const nextGrant =
    ordered.length === 0 ? [] : ordered.length === paths.length ? FULL_MODULE : ordered;

  return {
    ...current,
    [role]: {
      ...current[role],
      [moduleId]: nextGrant,
    },
  };
}

export function roleAllowsStaffPath(staffRoleAccess, pathname) {
  const match = findStaffNavCatalogItem(pathname);
  if (!match) return true;
  if (!staffRoleAccess || typeof staffRoleAccess !== "object") return true;
  return isRoleNavItemEnabled(staffRoleAccess, match.moduleId, match.item.path);
}

export function navItemAllowed(item, enabledList, staffRoleAccess = null) {
  if (!item.module && !item.modules?.length) return true;
  if (item.modules?.length) {
    if (!item.modules.some((m) => enabledList.includes(m))) return false;
  } else if (item.module && !enabledList.includes(item.module)) {
    return false;
  }
  if (!staffRoleAccess || !item.path) return true;
  if (!item.module) return true;
  return isRoleNavItemEnabled(staffRoleAccess, item.module, item.path);
}

export function filterNavGroupsByModules(groups, enabledList, staffRoleAccess = null) {
  return groups
    .map((group) => ({
      ...group,
      items: group.items.filter((item) => navItemAllowed(item, enabledList, staffRoleAccess)),
    }))
    .filter((group) => group.items.length > 0);
}

function pathMatches(pathname, prefixes) {
  return prefixes.some((p) => pathname === p || pathname.startsWith(`${p}/`));
}

function isPathAllowed(pathname, enabledList, always, rules) {
  if (!pathname) return true;
  if (pathMatches(pathname, always)) return true;
  const rule = rules.find((r) => pathMatches(pathname, r.prefixes));
  if (!rule) return true;
  return rule.modules.some((m) => enabledList.includes(m));
}

export function isStudentPathAllowed(pathname, enabled) {
  const list = normalizeEnabledModules(enabled).student;
  return isPathAllowed(pathname, list, STUDENT_ALWAYS, STUDENT_PATH_MODULES);
}

export function isStaffPathAllowed(pathname, enabled, staffRoleAccess = null) {
  const list = normalizeEnabledModules(enabled).staff;
  if (pathname === "/staff" || pathname === "/staff/") return true;
  if (!isPathAllowed(pathname, list, STAFF_ALWAYS, STAFF_PATH_MODULES)) return false;
  return roleAllowsStaffPath(staffRoleAccess, pathname);
}

export function isInstitutionAdminPathAllowed(pathname, enabled) {
  const list = normalizeEnabledModules(enabled).staff;
  if (pathname === "/institution/admin" || pathname === "/institution/admin/") return true;
  return isPathAllowed(pathname, list, INSTITUTION_ADMIN_ALWAYS, INSTITUTION_ADMIN_PATH_MODULES);
}

export function staffHomePath(user) {
  const enabled = normalizeEnabledModules(user?.enabled_modules).staff;
  if (user?.role === "scholarship_office") {
    if (enabled.includes("scholarships")) return "/staff/financial-aid";
    if (enabled.includes("grants")) return "/staff/grants";
    return "/staff";
  }
  return "/staff";
}

export function toggleModuleSelection(raw, portal, moduleId) {
  const current = normalizeEnabledModules(raw);
  const allowed = portal === "student"
    ? DEFAULT_ENABLED_MODULES.student
    : DEFAULT_ENABLED_MODULES.staff;
  if (!allowed.includes(moduleId)) return current;
  const set = new Set(current[portal]);
  if (set.has(moduleId)) set.delete(moduleId);
  else set.add(moduleId);
  return { ...current, [portal]: allowed.filter((id) => set.has(id)) };
}
